### PR TITLE
2057 extraction for nemo

### PR DESCRIPTION
**Pull request by hiker**
This adds standalone and NetCDF extraction libraries for NEMO, plus a runnable example that extracts data.

Driver creation is done separately in #2058.

### CODE REVIEW

**Comment by hiker** [comment]
The link check failures are expected, I have added links to the new library files, which do not yet exist. That should work once this is merged.

File: examples/nemo/README.md
Comment (arporter) [inline examples/nemo/README.md:141]: Ooh!

File: examples/nemo/eg5/Makefile
Comment (arporter) [inline examples/nemo/eg5/Makefile:4]: 2023?

File: examples/nemo/eg5/Makefile
Comment (arporter) [inline examples/nemo/eg5/Makefile:39]: "instruments each of the kernels"

File: examples/nemo/eg5/Makefile
Comment (arporter) [inline examples/nemo/eg5/Makefile:53]: I don't think you need traldf_iso?

File: examples/nemo/eg5/Makefile
Comment (arporter) [inline examples/nemo/eg5/Makefile:55]: This example doesn't use OMP.

File: examples/nemo/eg5/Makefile
Comment (arporter) [inline examples/nemo/eg5/Makefile:50]: Please add a comment here.

File: examples/nemo/eg5/Makefile
Comment (arporter) [inline examples/nemo/eg5/Makefile:81]: Good idea having the env. vars set here.

File: examples/nemo/eg5/Makefile
Comment (arporter) [inline examples/nemo/eg5/Makefile:89]: Since this is NEMO you don't need the -nodm flag and associated comment.

File: examples/nemo/eg5/README.md
Comment (arporter) [inline examples/nemo/eg5/README.md:20]: As in the Makefile, the NEMO API doesn't have explicit support DM so you don't need '-nodm'.

File: examples/nemo/eg5/Makefile
Comment (arporter) [reply examples/nemo/eg5/Makefile:81]: However, you need to update the name of the executable to include `${TYPE}` otherwise `TYPE=netcdf make run` doesn't work.

File: examples/nemo/eg5/domain_setup.sh
Comment (arporter) [inline examples/nemo/eg5/domain_setup.sh:4]: -2023

File: examples/nemo/eg5/extract_kernels.py
Comment (arporter) [inline examples/nemo/eg5/extract_kernels.py:36]: This doc string needs updating :-)

File: examples/nemo/eg5/extract_kernels.py
Comment (arporter) [inline examples/nemo/eg5/extract_kernels.py:76]: Please update.

File: examples/nemo/eg5/extract_kernels.py
Comment (arporter) [inline examples/nemo/eg5/extract_kernels.py:101]: Please add a comment explaining what the "COLLECT-ARRAY-SHAPE-READS" option setting is for.

File: lib/extract/netcdf/extract_netcdf_base.jinja
Comment (arporter) [inline lib/extract/netcdf/extract_netcdf_base.jinja:222]: What's the NF90_NETCDF4 for?

File: lib/extract/netcdf/nemo/kernel_data_netcdf.f90
Comment (arporter) [inline lib/extract/netcdf/nemo/kernel_data_netcdf.f90:39]: s/to around/around/

File: examples/nemo/README.md
Comment (arporter) [reply examples/nemo/README.md:141]: "this example"

Comment (hiker) [reply examples/nemo/README.md]: Done.

File: examples/nemo/eg5/Makefile
Comment (hiker) [reply examples/nemo/eg5/Makefile]: Indeed.

File: examples/nemo/eg5/Makefile
Comment (hiker) [reply examples/nemo/eg5/Makefile]: Done.

File: examples/nemo/eg5/Makefile
Comment (hiker) [reply examples/nemo/eg5/Makefile]: OOps ... and neither do I need psyclone.log.

File: examples/nemo/eg5/Makefile
Comment (hiker) [reply examples/nemo/eg5/Makefile]: Fixed.

File: examples/nemo/eg5/Makefile
Comment (hiker) [reply examples/nemo/eg5/Makefile:58]: Done.

File: examples/nemo/eg5/Makefile
Comment (hiker) [reply examples/nemo/eg5/Makefile]: Done.

File: examples/nemo/eg5/Makefile
Comment (hiker) [reply examples/nemo/eg5/Makefile:93]: Well, I actually do need it ... but I totally agree that this is stupid. So I've fixed this, the distributed memory flag is now only tested if the API is not NEMO. I've also added tests to see it works as expected with LFRic/GOcean, and a completely new test for nemo. I didn't add any more nemo specific tests for nemo - I expect that Nemo will always work with the base class (since it's pure PSyIR). At this stage I can't decide if we really want a new `nemo_extract_trans.py` file that just adds this one test (the test for DM is actually important at this stage for the extraction libraries, they do not support MPI). We would then have tests for distributed memory duplicated in any API-specific implementation, which is ugly as well. I suggest we leave it till the extraction libraries support MPI (TODO #1992), then it will just go away.

File: examples/nemo/eg5/README.md
Comment (hiker) [reply examples/nemo/eg5/README.md]: Removed.

File: examples/nemo/eg5/domain_setup.sh
Comment (hiker) [reply examples/nemo/eg5/domain_setup.sh]: Done.

File: examples/nemo/eg5/extract_kernels.py
Comment (hiker) [reply examples/nemo/eg5/extract_kernels.py]: Oops - totally.

File: examples/nemo/eg5/extract_kernels.py
Comment (hiker) [reply examples/nemo/eg5/extract_kernels.py]: Oops again.

File: examples/nemo/eg5/extract_kernels.py
Comment (hiker) [reply examples/nemo/eg5/extract_kernels.py:107]: Done - but tbh, we might need to look at a better option here.  FWIW, reason is the expression:
```
zwx(:,:,jpk) = 0.e0
```
which (I assume) is lowered to an expression using `shape(zwx, ..)` on the RHS, which then gives a problem with a combined read/write access on the left hand side. Of course, jf we should have code lik:
```
something_else(shape(zwx, 1),jpk) = 0.e0
```
then we do need the read access (or we get smart and only store the result of `shape`. But that will require a lot of additional work in the driver later, which I wouldn't like ... )? I'd guess converting array expressions to explicit loops would also solve this issue.

File: lib/extract/netcdf/extract_netcdf_base.jinja
Comment (hiker) [reply lib/extract/netcdf/extract_netcdf_base.jinja:223]: I've added a comment (we need netcdf 4 Format in order to use 64-bit integer in NF90_INT_64 or so).

File: lib/extract/netcdf/nemo/kernel_data_netcdf.f90
Comment (hiker) [reply lib/extract/netcdf/nemo/kernel_data_netcdf.f90]: Done.

**Comment by hiker** [comment]
As before, the link check errors are expected, since I've added new locations which do not yet exist.

File: examples/nemo/eg5/extract_kernels.py
Comment (arporter) [inline examples/nemo/eg5/extract_kernels.py:101]: "PSyclone represents this internally using `Range` with `LBOUND` and `UBOUND` intrinsics and currently this results in several..."

File: lib/extract/netcdf/extract_netcdf_base.jinja
Comment (arporter) [reply lib/extract/netcdf/extract_netcdf_base.jinja:223]: Great, thanks.

**Review by arporter — CHANGES_REQUESTED** [review]
Very nearly there now.
Just a couple of corrections needed.
Also, I realised the UG section on examples needs updating (file:///home/kbc59144/Projects/PSyclone/doc/user_guide/_build/html/examples.html#nemo).
All tests and examples are OK with compilation.
Ref. guide builds fine.

File: examples/nemo/eg5/Makefile
Comment (hiker) [reply examples/nemo/eg5/Makefile]: Oops.

File: examples/nemo/eg5/extract_kernels.py
Comment (hiker) [reply examples/nemo/eg5/extract_kernels.py]: Done.

**Comment by hiker** [comment]
I've updated #2080 to look at using `is_full_access` (I had a look, that should be simple, just a bit too much to sneak this in as part of this PR :) ), and then added a TODO for #2080 to the example, so we can update the comment.
Ready for next review.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
UG has been updated and builds fine.
Examples are still all OK with compilation.
Will proceed to merge.

**Pull request merged**
